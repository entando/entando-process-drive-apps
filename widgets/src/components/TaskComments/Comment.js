import React from 'react';
import PropTypes from 'prop-types';
import i18next from 'i18next';

import Avatar from '@material-ui/core/Avatar';
import withStyles from '@material-ui/core/styles/withStyles';

const styles = {
  comment: {
    display: 'flex',
    padding: '5px 10px',
    margin: '10px 0px',
  },
  avatar: {
    margin: '0px 30px',
  },
  commentText: {
    fontSize: '13px',
    lineHeight: '15px',
    color: '#686C6E',
  },
  created: {
    fontSize: '10px',
    lineHeight: '12px',
    color: '#8B8888',
    marginTop: '10px',
    marginLeft: '7px',
  },
  removeAnswers: {
    marginLeft: '5px',
    marginRight: '5px',
  },
};

class Comment extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      prompt: false,
    };

    this.togglePrompt = this.togglePrompt.bind(this);
    this.handleClickRemove = this.handleClickRemove.bind(this);
  }

  togglePrompt() {
    const { prompt } = this.state;
    this.setState({ prompt: !prompt });
  }

  handleClickRemove() {
    const {
      comment: { id },
      onClickRemoveComment,
    } = this.props;

    onClickRemoveComment(id);
  }

  render() {
    const { prompt } = this.state;
    const {
      classes,
      comment: { text, createdBy, createdAt },
    } = this.props;

    return (
      <div className={classes.comment}>
        <div className={classes.avatar}>
          <Avatar alt={createdBy} src="/" />
        </div>
        <div className={classes.commentText}>
          <div>{text}</div>
          <div className={classes.created}>
            {`${createdBy} - ${new Date(createdAt).toLocaleString(i18next.language || 'en')} - `}
            <span
              onClick={this.togglePrompt}
              onKeyPress={this.togglePrompt}
              role="button"
              tabIndex={0}
            >
              {i18next.t('taskComments.remove')}
            </span>
            {prompt && (
              <span>
                <span
                  onClick={this.handleClickRemove}
                  onKeyPress={this.handleClickRemove}
                  role="button"
                  tabIndex={0}
                  className={classes.removeAnswers}
                >
                  {i18next.t('common.yes')}
                </span>
                /
                <span
                  onClick={this.togglePrompt}
                  onKeyPress={this.togglePrompt}
                  role="button"
                  tabIndex={0}
                  className={classes.removeAnswers}
                >
                  {i18next.t('common.no')}
                </span>
              </span>
            )}
          </div>
        </div>
      </div>
    );
  }
}

Comment.propTypes = {
  classes: PropTypes.shape({
    comment: PropTypes.string,
    avatar: PropTypes.string,
    commentText: PropTypes.string,
    created: PropTypes.string,
    removeAnswers: PropTypes.string,
  }).isRequired,
  comment: PropTypes.shape({
    id: PropTypes.string,
    text: PropTypes.string,
    createdBy: PropTypes.string,
    createdAt: PropTypes.string,
  }).isRequired,
  onClickRemoveComment: PropTypes.func,
};

Comment.defaultProps = {
  onClickRemoveComment: () => {},
};

export default withStyles(styles)(Comment);
