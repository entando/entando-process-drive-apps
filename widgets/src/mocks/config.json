{
  "payload": {
    "code": "pda_task_list",
    "config": {
      "groups": "[{\"label\":\"Administrators\",\"key\":\"Administrators\",\"checked\":true},{\"label\":\"broker\",\"key\":\"broker\",\"checked\":true},{\"label\":\"admin\",\"key\":\"admin\",\"checked\":false},{\"label\":\"rest-all\",\"key\":\"rest-all\",\"checked\":false},{\"label\":\"kiemgmt\",\"key\":\"kiemgmt\",\"checked\":true},{\"label\":\"kie-server\",\"key\":\"kie-server\",\"checked\":true},{\"label\":\"appraiser\",\"key\":\"appraiser\",\"checked\":false},{\"label\":\"manager\",\"key\":\"manager\",\"checked\":true},{\"label\":\"supplier\",\"key\":\"supplier\",\"checked\":true},{\"label\":\"IT\",\"key\":\"IT\",\"checked\":true},{\"label\":\"PM\",\"key\":\"PM\",\"checked\":true}]",
      "options": "[{\"key\":\"newPageOnClick\",\"label\":\"Open new page on table row click\",\"checked\":false},{\"key\":\"showClaim\",\"label\":\"Show Claim Button\",\"checked\":true},{\"key\":\"showComplete\",\"label\":\"Show Complete Button\",\"checked\":true}]",
      "columns": "[{\"name\":\"task-is-skipable\",\"position\":0,\"isVisible\":false},{\"name\":\"task-parent-id\",\"position\":1,\"isVisible\":false},{\"name\":\"task-created-by\",\"position\":2,\"isVisible\":true},{\"name\":\"task-input-data.application.property.saleprice\",\"position\":3,\"isVisible\":true},{\"name\":\"task-input-data.TaskName\",\"position\":4,\"isVisible\":true},{\"name\":\"initiator\",\"position\":5,\"isVisible\":true},{\"name\":\"task-created-on\",\"position\":6,\"isVisible\":true},{\"name\":\"task-priority\",\"position\":7,\"isVisible\":false},{\"name\":\"task-proc-def-id\",\"position\":8,\"isVisible\":true},{\"name\":\"task-form\",\"position\":9,\"isVisible\":true},{\"name\":\"task-skippable\",\"position\":10,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.name\",\"position\":11,\"isVisible\":true},{\"name\":\"task-actual-owner\",\"position\":12,\"isVisible\":true},{\"name\":\"task-activation-time\",\"position\":13,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.address\",\"position\":14,\"isVisible\":true},{\"name\":\"task-input-data.application.errors\",\"position\":15,\"isVisible\":true},{\"name\":\"task-name\",\"position\":16,\"isVisible\":true},{\"name\":\"task-input-data.application.property.address\",\"position\":17,\"isVisible\":true},{\"name\":\"task-subject\",\"position\":18,\"isVisible\":true},{\"name\":\"containerId\",\"position\":19,\"isVisible\":true},{\"name\":\"task-container-id\",\"position\":20,\"isVisible\":true},{\"name\":\"processDefinitionId\",\"position\":21,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.ssn\",\"position\":22,\"isVisible\":true},{\"name\":\"task-description\",\"position\":23,\"isVisible\":true},{\"name\":\"task-input-data.Skippable\",\"position\":24,\"isVisible\":true},{\"name\":\"task-workitem-id\",\"position\":25,\"isVisible\":true},{\"name\":\"task-expiration-time\",\"position\":26,\"isVisible\":true},{\"name\":\"task-input-data.application.property.age\",\"position\":27,\"isVisible\":true},{\"name\":\"task-status\",\"position\":28,\"isVisible\":true},{\"name\":\"task-input-data.application.mortgageamount\",\"position\":29,\"isVisible\":true},{\"name\":\"task-process-instance-id\",\"position\":30,\"isVisible\":true},{\"name\":\"task-input-data.ActorId\",\"position\":31,\"isVisible\":true},{\"name\":\"task-process-id\",\"position\":32,\"isVisible\":true},{\"name\":\"task-input-data.application.property.locale\",\"position\":33,\"isVisible\":true},{\"name\":\"task-input-data.application.amortization\",\"position\":34,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.creditrating\",\"position\":35,\"isVisible\":true},{\"name\":\"application\",\"position\":36,\"isVisible\":true},{\"name\":\"task-input-data.NodeName\",\"position\":37,\"isVisible\":true},{\"name\":\"task-proc-inst-id\",\"position\":38,\"isVisible\":true},{\"name\":\"task-input-data.application.downpayment\",\"position\":39,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.annualincome\",\"position\":40,\"isVisible\":true},{\"name\":\"task-id\",\"position\":41,\"isVisible\":true}]",
      "knowledgeSource": "kieStaging",
      "process": "itorders.orderhardware@itorders_1.0.0-SNAPSHOT"
    }
  },
  "metaData": {
    "status": "draft"
  },
  "errors": []
}
