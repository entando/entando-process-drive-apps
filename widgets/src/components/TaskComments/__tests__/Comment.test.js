import { render, fireEvent, wait } from '@testing-library/react';
import React from 'react';
import 'mocks/i18nMock';

import Comment from 'components/TaskComments/Comment';

import MOCKED_GET_TASK_COMMENTS_RESPONSE from 'mocks/taskComments/getComments';

describe('<Comment />', () => {
  const MOCKED_COMMENT = MOCKED_GET_TASK_COMMENTS_RESPONSE.payload[0];

  const fixedDate = new Date(2020, 0, 1);
  global.Date = jest.fn(() => fixedDate);
  global.Date.now = jest.fn(() => fixedDate);

  it('renders snapshot correctly', async () => {
    const { container } = render(<Comment comment={MOCKED_COMMENT} />);

    expect(container).toMatchSnapshot();
  });

  it('shows Yes/No prompt when "Remove" is pressed', async () => {
    const { findByText, queryByText } = render(<Comment comment={MOCKED_COMMENT} />);

    const removeButton = await findByText('taskComments.remove');
    fireEvent.click(removeButton);

    await wait(() => {
      expect(queryByText('common.yes')).toBeInTheDocument();
      expect(queryByText('common.no')).toBeInTheDocument();
    });
  });

  it('hides prompt when No is pressed', async () => {
    const { findByText, queryByText } = render(<Comment comment={MOCKED_COMMENT} />);

    const removeButton = await findByText('taskComments.remove');
    fireEvent.click(removeButton);

    const noButton = await findByText('common.no');
    fireEvent.click(noButton);

    await wait(() => {
      expect(queryByText('common.yes')).not.toBeInTheDocument();
      expect(queryByText('common.no')).not.toBeInTheDocument();
    });
  });

  it('calls onClickRemoveComment function when Remove -> Yes is pressed', async () => {
    const onClickRemoveComment = jest.fn();
    const { findByText } = render(
      <Comment comment={MOCKED_COMMENT} onClickRemoveComment={onClickRemoveComment} />
    );

    const removeButton = await findByText('taskComments.remove');
    fireEvent.click(removeButton);

    const yesButton = await findByText('common.yes');
    fireEvent.click(yesButton);

    await wait(() => {
      expect(onClickRemoveComment).toHaveBeenCalledTimes(1);
    });
  });
});
