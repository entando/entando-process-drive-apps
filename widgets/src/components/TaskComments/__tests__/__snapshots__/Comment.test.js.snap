// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comment /> renders snapshot correctly 1`] = `
<div>
  <div
    class="Comment-comment-1"
  >
    <div
      class="Comment-avatar-2"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circle"
      >
        <img
          alt="pamAdmin"
          class="MuiAvatar-img"
          src="/"
        />
      </div>
    </div>
    <div
      class="Comment-commentText-3"
    >
      <div>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis volutpat est ipsum, quis congue est pharetra ut. Morbi mollis sem eu tellus elementum, et bibendum dolor viverra. Donec maximus quam ligula, et commodo risus ultricies vel. Morbi mollis dui odio, id accumsan turpis pulvinar vitae. Mauris fringilla in mauris nec ullamcorper. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Sed eu sagittis metus, vitae posuere neque. Etiam ultrices, leo a convallis pharetra, magna ante mollis est, ut tristique neque mauris quis lacus.

In pulvinar, leo quis congue molestie, metus dolor sagittis magna, nec efficitur leo tortor at orci. Sed fringilla urna non urna dapibus, ac imperdiet libero condimentum. Phasellus vulputate vel est eu facilisis. Pellentesque eu purus egestas, condimentum ante finibus, consectetur purus. Sed commodo ante aliquam, vestibulum est at, laoreet lorem. Etiam blandit mi sed lectus fermentum, sit amet lacinia ligula egestas. Curabitur id elit mattis, ultricies mauris sagittis, viverra nulla. Nunc id ligula a dolor placerat porttitor ut in neque. Curabitur imperdiet felis a magna fringilla, nec tempus justo scelerisque.
      </div>
      <div
        class="Comment-created-4"
      >
        pamAdmin - 1/29/2020, 11:26:02 AM - 
        <span
          role="button"
          tabindex="0"
        >
          taskComments.remove
        </span>
      </div>
    </div>
  </div>
</div>
`;
